File: presto-main-base/src/main/java/com/facebook/presto/operator/scalar/KllSketchFunctions.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.facebook.presto.operator.scalar;

import com.facebook.presto.spi.function.Description;
import com.facebook.presto.spi.function.ScalarFunction;
import com.facebook.presto.spi.function.SqlType;
import io.airlift.slice.Slice;
import io.airlift.slice.Slices;
import org.apache.datasketches.common.ArrayOfBooleansSerDe;
import org.apache.datasketches.common.ArrayOfDoublesSerDe;
import org.apache.datasketches.common.ArrayOfLongsSerDe;
import org.apache.datasketches.common.ArrayOfStringsSerDe;
import org.apache.datasketches.kll.KllItemsSketch;
import org.apache.datasketches.memory.Memory;
import org.apache.datasketches.quantilescommon.QuantileSearchCriteria;

import static com.facebook.presto.common.type.StandardTypes.BIGINT;
import static com.facebook.presto.common.type.StandardTypes.BOOLEAN;
import static com.facebook.presto.common.type.StandardTypes.DOUBLE;
import static com.facebook.presto.common.type.StandardTypes.VARCHAR;
import static java.nio.ByteOrder.LITTLE_ENDIAN;

public class KllSketchFunctions
{
    private KllSketchFunctions()
    {
    }

    @ScalarFunction("sketch_kll_quantile")
    @Description("Calculates the quantile for a given value with the provided inclusivity. If inclusive is true, the given rank includes all quantiles ≤ the quantile directly corresponding to the given rank. If false, the given rank includes all quantiles < the quantile directly corresponding to the given rank.")
    @SqlType(DOUBLE)
    public static double sketchQuantileDouble(
            @SqlType("kllsketch(double)") Slice rawSketch,
            @SqlType(DOUBLE) double rank,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<Double> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), Double::compareTo, new ArrayOfDoublesSerDe());
        return sketch.getQuantile(rank, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(DOUBLE)
    public static double sketchQuantileDouble(
            @SqlType("kllsketch(double)") Slice rawSketch,
            @SqlType(DOUBLE) double rank)

    {
        return sketchQuantileDouble(rawSketch, rank, true);
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(BIGINT)
    public static long sketchQuantileBigint(
            @SqlType("kllsketch(bigint)") Slice rawSketch,
            @SqlType(DOUBLE) double rank,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<Long> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), Long::compareTo, new ArrayOfLongsSerDe());
        return sketch.getQuantile(rank, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(BIGINT)
    public static long sketchQuantileBigint(
            @SqlType("kllsketch(bigint)") Slice rawSketch,
            @SqlType(DOUBLE) double rank)

    {
        return sketchQuantileBigint(rawSketch, rank, true);
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(VARCHAR)
    public static Slice sketchQuantileString(
            @SqlType("kllsketch(varchar)") Slice rawSketch,
            @SqlType(DOUBLE) double rank,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<String> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), String::compareTo, new ArrayOfStringsSerDe());
        return Slices.utf8Slice(sketch.getQuantile(rank, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE));
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(VARCHAR)
    public static Slice sketchQuantileString(
            @SqlType("kllsketch(varchar)") Slice rawSketch,
            @SqlType(DOUBLE) double rank)

    {
        return sketchQuantileString(rawSketch, rank, true);
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(BOOLEAN)
    public static boolean sketchQuantileBoolean(
            @SqlType("kllsketch(boolean)") Slice rawSketch,
            @SqlType(DOUBLE) double rank,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<Boolean> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), Boolean::compareTo, new ArrayOfBooleansSerDe());
        return sketch.getQuantile(rank, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_quantile")
    @SqlType(BOOLEAN)
    public static boolean sketchQuantileBoolean(
            @SqlType("kllsketch(boolean)") Slice rawSketch,
            @SqlType(DOUBLE) double rank)

    {
        return sketchQuantileBoolean(rawSketch, rank, true);
    }

    @ScalarFunction("sketch_kll_rank")
    @Description("Calculates the rank of a quantile; An estimate of the value which occurs at a particular quantile in the distribution")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(double)") Slice rawSketch,
            @SqlType(DOUBLE) double quantile,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<Double> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), Double::compareTo, new ArrayOfDoublesSerDe());
        return sketch.getRank(quantile, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(double)") Slice rawSketch,
            @SqlType(DOUBLE) double quantile)

    {
        return sketchRank(rawSketch, quantile, true);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(bigint)") Slice rawSketch,
            @SqlType(BIGINT) long quantile,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<Long> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), Long::compareTo, new ArrayOfLongsSerDe());
        return sketch.getRank(quantile, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(bigint)") Slice rawSketch,
            @SqlType(BIGINT) long quantile)

    {
        return sketchRank(rawSketch, quantile, true);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(varchar)") Slice rawSketch,
            @SqlType(VARCHAR) Slice rank,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<String> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), String::compareTo, new ArrayOfStringsSerDe());
        return sketch.getRank(rank.toStringUtf8(), inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(varchar)") Slice rawSketch,
            @SqlType(VARCHAR) Slice quantile)

    {
        return sketchRank(rawSketch, quantile, true);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(boolean)") Slice rawSketch,
            @SqlType(BOOLEAN) boolean rank,
            @SqlType(BOOLEAN) boolean inclusive)

    {
        KllItemsSketch<Boolean> sketch = KllItemsSketch.wrap(Memory.wrap(rawSketch.toByteBuffer(), LITTLE_ENDIAN), Boolean::compareTo, new ArrayOfBooleansSerDe());
        return sketch.getRank(rank, inclusive ? QuantileSearchCriteria.INCLUSIVE : QuantileSearchCriteria.EXCLUSIVE);
    }

    @ScalarFunction("sketch_kll_rank")
    @SqlType(DOUBLE)
    public static double sketchRank(
            @SqlType("kllsketch(boolean)") Slice rawSketch,
            @SqlType(BOOLEAN) boolean quantile)

    {
        return sketchRank(rawSketch, quantile, true);
    }
}

File: presto-native-execution/presto_cpp/main/functions/kll_sketch/KllSketchFunctions.cpp
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "DataSketches/kll_sketch.hpp"

#include "presto_cpp/main/functions/kll_sketch/KllSketchRegistration.h"
#include "presto_cpp/main/types/KllSketchType.h"
#include "velox/velox/functions/Macros.h"
#include "velox/velox/functions/Registerer.h"

namespace facebook::presto::functions {

namespace {

//---------------------------------------------
// Type mapping: Velox types -> DataSketches types
//---------------------------------------------

template <typename T>
struct SketchTypeMapper {
  using type = T;
};

// VARCHAR in Velox maps to std::string in DataSketches
template <>
struct SketchTypeMapper<velox::Varchar> {
  using type = std::string;
};

template <typename T>
using SketchType = typename SketchTypeMapper<T>::type;

//---------------------------------------------
// Type conversion helpers
//---------------------------------------------

template <typename T>
inline SketchType<T> convertValue(const T& value) {
  return value;
}

// Velox VARCHAR → std::string for DataSketches
inline std::string convertValue(const velox::StringView& value) {
  return std::string(value.data(), value.size());
}

//---------------------------------------------
// Sketch deserialization helper
//---------------------------------------------

template <typename T>
inline datasketches::kll_sketch<SketchType<T>> deserializeSketch(
    const velox::StringView& rawSketch) {
  try {
    return datasketches::kll_sketch<SketchType<T>>::deserialize(
        rawSketch.data(), rawSketch.size());
  } catch (const std::out_of_range& e) {
    // Thrown by DataSketches for insufficient buffer size
    VELOX_USER_FAIL(
        "Invalid KLL sketch data - buffer out of range: {}", e.what());
  } catch (const std::logic_error& e) {
    // Thrown by DataSketches for corrupted data (size mismatch, etc.)
    VELOX_USER_FAIL(
        "Failed to deserialize KLL sketch - corrupted data or logic error: {}",
        e.what());
  } catch (const std::exception& e) {
    // Catch any other unexpected exceptions during deserialization
    VELOX_USER_FAIL("Failed to deserialize KLL sketch: {}", e.what());
  }
}

//---------------------------------------------
// Rank Function (supports both 2 and 3 args)
//---------------------------------------------

template <typename T>
struct KllSketchRankFunction {
  template <typename TExec>
  struct udf {
    VELOX_DEFINE_FUNCTION_TYPES(TExec);

    FOLLY_ALWAYS_INLINE bool callNullable(
        out_type<double>& result,
        const arg_type<velox::Varbinary>* rawSketch,
        const arg_type<T>* quantile,
        const arg_type<bool>* inclusive) {
      if (!rawSketch || !quantile || !inclusive) {
        return false;
      }

      auto sketch = deserializeSketch<T>(*rawSketch);
      auto value = convertValue(*quantile);

      result = sketch.get_rank(value, *inclusive);
      return true;
    }

    FOLLY_ALWAYS_INLINE bool callNullable(
        out_type<double>& result,
        const arg_type<velox::Varbinary>* rawSketch,
        const arg_type<T>* quantile) {
      if (!rawSketch || !quantile) {
        return false;
      }

      auto sketch = deserializeSketch<T>(*rawSketch);
      auto value = convertValue(*quantile);

      result = sketch.get_rank(value, true);
      return true;
    }
  };
};

//---------------------------------------------
// Quantile Function (supports both 2 and 3 args)
//---------------------------------------------

template <typename T>
struct KllSketchQuantileFunction {
  template <typename TExec>
  struct udf {
    VELOX_DEFINE_FUNCTION_TYPES(TExec);

    // Helper to write result (handles VARCHAR differently)
    FOLLY_ALWAYS_INLINE void writeResult(
        out_type<T>& result,
        const SketchType<T>& value) {
      if constexpr (std::is_same_v<T, velox::Varchar>) {
        result.append(value);
      } else {
        result = value;
      }
    }

    FOLLY_ALWAYS_INLINE bool callNullable(
        out_type<T>& result,
        const arg_type<velox::Varbinary>* rawSketch,
        const arg_type<double>* rank,
        const arg_type<bool>* inclusive) {
      if (!rawSketch || !rank || !inclusive) {
        return false;
      }

      auto sketch = deserializeSketch<T>(*rawSketch);

      auto quantile = sketch.get_quantile(*rank, *inclusive);
      writeResult(result, quantile);
      return true;
    }

    FOLLY_ALWAYS_INLINE bool callNullable(
        out_type<T>& result,
        const arg_type<velox::Varbinary>* rawSketch,
        const arg_type<double>* rank) {
      if (!rawSketch || !rank) {
        return false;
      }

      auto sketch = deserializeSketch<T>(*rawSketch);

      auto quantile = sketch.get_quantile(*rank, true);
      writeResult(result, quantile);
      return true;
    }
  };
};

} // namespace

//---------------------------------------------
// Registration Helper
//---------------------------------------------

template <typename T, typename SketchArg>
void registerKllType(
    const std::string& rankFuncName,
    const std::string& quantileFuncName) {
  // Register rank with 3 args (sketch, value, inclusive)
  velox::registerFunction<
      KllSketchRankFunction<T>,
      double,
      SimpleKllSketch<SketchArg>,
      T,
      bool>({rankFuncName});

  // Register rank with 2 args (sketch, value) - defaults inclusive=true
  velox::registerFunction<
      KllSketchRankFunction<T>,
      double,
      SimpleKllSketch<SketchArg>,
      T>({rankFuncName});

  // Register quantile with 3 args (sketch, rank, inclusive)
  velox::registerFunction<
      KllSketchQuantileFunction<T>,
      T,
      SimpleKllSketch<SketchArg>,
      double,
      bool>({quantileFuncName});

  // Register quantile with 2 args (sketch, rank) - defaults inclusive=true
  velox::registerFunction<
      KllSketchQuantileFunction<T>,
      T,
      SimpleKllSketch<SketchArg>,
      double>({quantileFuncName});
}

//---------------------------------------------
// Function Registration
//---------------------------------------------

void registerKllSketchFunctions(const std::string& prefix) {
  const std::string rankFuncName = prefix + "sketch_kll_rank";
  const std::string quantileFuncName = prefix + "sketch_kll_quantile";

  // Register for all supported types
  registerKllType<double, double>(rankFuncName, quantileFuncName);
  registerKllType<int64_t, int64_t>(rankFuncName, quantileFuncName);
  registerKllType<bool, bool>(rankFuncName, quantileFuncName);

  // VARCHAR special case: Velox uses StringView, DataSketches uses std::string
  registerKllType<velox::Varchar, velox::StringView>(
      rankFuncName, quantileFuncName);
}

} // namespace facebook::presto::functions
